### PR TITLE
[SE-0286] Forward matching of trailing closure arguments

### PR DESCRIPTION
**Pull request by DougGregor**
Introduce a new "forward" algorithm for trailing closures where
the unlabeled trailing closure argument matches the next parameter in
the parameter list that can accept an unlabeled trailing closure. This
feature is described by SE-0286.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci build toolchain

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -5451,12 +5447,201 @@ static bool hasCurriedSelf(ConstraintSystem &cs, ConcreteDeclRef callee,
...
+    auto decl = callee.getDecl();
+    if (!decl)
+      return;
+
+    auto diag = ctx.Diags.diagnose(
+        decl, diag::decl_multiple_defaulted_closure_parameters,
+        decl->getName(), paramName, backwardParamName);
+
+    // Dig out the parameter declarations so we can highlight them.
+    // FIXME: There should be a utility for this.
```
Comment (theblixguy) [inline lib/Sema/CSApply.cpp:5557]: You can use `swift::getParameterList` :)

Comment (DougGregor) [reply lib/Sema/CSApply.cpp]: Fantastic, thank you! I've switched over to using that here.

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please build toolchain

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci test platform Linux

**Comment by DougGregor** [comment]
@swift-ci test platform Windows

**Comment by DougGregor** [comment]
@swift-ci smoke test Linux

**Comment by DougGregor** [comment]
@swift smoke test

**Pull request closed without merging**